Consolidate nondeterministic error tests

kurtamohler opened the pull request:
Fixes #51498

kurtamohler committed:
Consolidate nondeterministic error tests

kurtamohler commented on test/test_torch.py:
```diff
@@ -2905,6 +2905,56 @@ def test_use_deterministic_algorithms_beta_warning(self, device):
+            ),
+            (
+                'index_copy',
+                (lambda slf, device: torch.randn(10, device=device).index_copy(0, torch.tensor([3], device=device), torch.randn(1, device=device))),
```
This might not be a very good design. If the test function needs to be multiple lines, then it will need to be defined up above. Eventually, we might have many test functions all packed into `test_use_deterministic_algorithms_errors`. That's not necessarily wrong, but it could get difficult to keep track of indentations.

mruberry replied:
This is a cool design for a tricky problem.

Would it just be simpler to create two tests: test_use_deterministic_algorithms_cpu_errors and test_use_deterministic_algorithms_cuda_errors and just manually create the input to trigger each determinism warning? The first test would have the @onlyCPU decorator and the second would have the @onlyCUDA decorator.

This alone would still be a little verbose because each op would need a line to create the input, a line to use the context manager to catch error, and a line to call the op. But that seems OK? Would that really be many more LoCs than trying a more structured approach like this one? 

Also happy to talk about ideas more on Slack.

kurtamohler replied:
Yeah I guess it's not too difficult to manage as it is.

I'm not sure I understand the benefit of splitting cpu and cuda into two different tests. Some errors happen for both devices, so I figure it's best to keep them combined

mruberry replied:
That's a good point. Sections of the test can also guard on the device type being CPU or CUDA if that's easier.

kurtamohler commented on test/test_torch.py:
```diff
@@ -2905,6 +2905,56 @@ def test_use_deterministic_algorithms_beta_warning(self, device):
+            ),
+        ]
+
+        for name, test_func, error_device, caller_name_forward, caller_name_backward in nondeterministic_operations:
```
Perhaps instead of looping through the test cases within a unit test, we could instead do something similar to `method_tests` and `OpInfo`, where we create a full unit test for each of these cases. That would make it easier to debug failures.

kurtamohler commented:
@mruberry, right now, this PR is a draft to nail down how we should organize these tests, before taking the time to move them all. Please let me know what you think about the current state, and any thoughts on my comments

kurtamohler commented on test/test_torch.py:
```diff
@@ -2906,6 +2906,413 @@ def test_use_deterministic_algorithms_beta_warning(self, device):
+        #
+        # caller_name_backward  Name of function shown in error for backward pass.
+        #                       Set to None if the backward should not throw an error.
+        nondeterministic_operations = [
```
@mruberry, I realized I could simplify these entries a little bit and clean up the long lines by adding a constructor function that defines the args and kwargs to give to the forward function. This has a bit of a weird consequence that the nn modules need to specify their `.__call__` method for the `operation` parameter. Maybe this is bad? I'm not sure. We could potentially just use the `nn.functional` version of all the nn modules to make it look a little nicer.

mruberry replied:
This looks OK to me. Whether we want to stick with this approach or just hand-write the "subtests" by creating the inputs, calling the op, and then verifying the error is thrown, is just a question of # LoC and readability/maintainability. For maintainability, consider that this will need to be documented in the pattern for adding a deterministic operation, and if people have questions about how to extend the test they'll come to you, so whatever minimizes those questions is probably the best thing to do.

kurtamohler replied:
Good points. Considering the challenges of making the structured approach flexible enough to fit every case and simple enough to understand easily, handwritten tests seem like a better way to go

kurtamohler commented on test/test_torch.py:
```diff
@@ -2906,6 +2906,413 @@ def test_use_deterministic_algorithms_beta_warning(self, device):
+                'cuda', 'kthvalue CUDA', None
+            ),
+            (
+                'median_indices',
```
TODO: Some of the current tests, like `test_median_nondeterministic_alert` lower in this file, test the function, method, and out variants. `test_median_nondeterministic_alert` also tests to make sure that when `indices` is not specified, it does not throw a deterministic error. We need to be able to incorporate these cases here

mruberry replied:
handwriting does make special cases easy to support, and a good "template" or "pattern" showing people how to handwrite these tests is comparable to structuring the pattern in code...

mruberry commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -862,8 +862,10 @@ def __init__(self, caller_name, device_type=None, fn_has_device_arg=True):
         def efail_fn(slf, device, *args, **kwargs):
-            if self.device_type is None or self.device_type == slf.device_type:
-                with DeterministicGuard(True):
+            with DeterministicGuard(True):
```
These changes look good, btw

kurtamohler committed:
Split nondeterministic alert test into multiple tests

kurtamohler committed:
Remove old tests, move remaining ones to test_torch.py

kurtamohler committed:
Fix docs for NLLLoss and grid_sample

kurtamohler committed:
Disable gather test for XLA

kurtamohler commented:
@mruberry, I've switched back to manually written tests. Now all of the tests for the nondeterministic alert are in `test/test_torch.py`. They use a fairly consistent pattern that shouldn't be too difficult for people to mimic for new tests. I added a point in `Note [Writing Nondeterministic Operations]` to try to enforce testing consistency in the future. There were also two docs mistakes that I fixed. I think this is ready for a review now

kurtamohler committed:
Fix broken link

kurtamohler committed:
Merge branch 'master' into consolidate-deterministic-tests

mruberry commented on torch/testing/_internal/common_nn.py:
```diff
@@ -2227,15 +2191,6 @@ def fractional_max_pool3d_test(test_case):
         skip_half=True,
         desc='complex'
     ),
-    dict(
```
Is it actually OK to remove these? Were they only added for nondeterministic testing, and are they only used for that purpose still?

kurtamohler replied:
I added most, if not all, of these just for the purpose of nondeterministic testing. But let me git blame all of them to make sure they weren't changed afterward for any other purpose.

kurtamohler replied:
Confirmed, these are all safe to be deleted

mruberry reviewed (commented):
Hey @kurtamohler! The new approach looks great. It's clear and simple and easy to extend. The other fixes look good, too.

I just have one question if we can be so aggressive about removing some of those module entries, or if some of them are still used by other tests.

mruberry reviewed (approved):
Cool!

kurtamohler committed:
Merge branch 'master' into consolidate-deterministic-tests
